libnavigation-core/src/main/java/com/mapbox/navigation/core/preview/RoutesPreviewController.kt

```diff
@@ -0,0 +1,116 @@
+package com.mapbox.navigation.core.preview
+
+import androidx.annotation.UiThread
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.core.routealternatives.mapToMetadata
+import com.mapbox.navigator.RoutesData
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.util.concurrent.CopyOnWriteArrayList
+
+@ExperimentalPreviewMapboxNavigationAPI
+@UiThread
+internal class RoutesPreviewController(
+    private val routesDataParser: RoutesDataParser,
+    private val scope: CoroutineScope
+) {
+
+    private val mutex = Mutex()
+
+    private var lastUpdate: RoutesPreviewUpdate = RoutesPreviewUpdate(
+        RoutesPreviewUpdateReasons.PREVIEW_CLEAN_UP,
+        null
+    )
+        set(value) {
+            if (value != field) {
+                field = value
+                observers.forEach {
+                    it.routesPreviewUpdated(value)
+                }
+            }
+        }
+
+    private val observers = CopyOnWriteArrayList<RoutesPreviewObserver>()
+
+    fun registerPreviewRoutesObserver(observer: RoutesPreviewObserver) {
+        observers.add(observer)
+        observer.routesPreviewUpdated(lastUpdate)
+    }
+
+    fun unregisterPreviewRoutesObserver(observer: RoutesPreviewObserver) {
+        observers.remove(observer)
+    }
+
+    fun getRoutesPreview(): RoutesPreview? = lastUpdate.routesPreview
+
+    fun previewNavigationRoutes(routesToPreview: List<NavigationRoute>) {
+        scope.launch {
+            mutex.withLock {
+                previewRoutesInternal(routesToPreview)
+            }
+        }
+    }
+
+    @Throws(IllegalArgumentException::class)
+    fun setRoutesPreviewPrimaryRoute(route: NavigationRoute) {
+        val originalRoutes = lastUpdate.routesPreview?.originalRoutesList
+        require(originalRoutes != null) {
+            "no previewed routes are set"
+        }
+        val routes = originalRoutes.toMutableList()
+        require(routes.remove(route)) {
+            "route $route isn't found among the list of previewed routes"
+        }
+        routes.add(0, route)
+        scope.launch {
```

Comment:
Shouldn't the mutex be used here? What if you invoke `setRoutesPreviewPrimaryRoute` twice in a short time? Or `setRoutesPreviewPrimaryRoute` and `previewNavigationRoutes `?  Is the order of updates guaranteed to be the same as the order of invocations?

Reply:
yet, it should! will add that, thank you!

Reply:
BTW, yes, these methods are asynchronous. Is it OK? I think that the caller would expect to get the updated previewed routes right after they are set.
Maybe yes, since we provide the data via an observer and `getRoutesPreview()` is just a snapshot of current state. We could also introduce a callback for the method but I'm not sure it's necessary as we already have an observer.